lib/Driver/WindowsToolChains.cpp

```diff
@@ -85,28 +85,14 @@ toolchains::Windows::constructInvocation(const DynamicLinkJobAction &job,
   }
 
   // Configure the toolchain.
-  //
-  // By default use the system `clang` to perform the link.  We use `clang` for
-  // the driver here because we do not wish to select a particular C++ runtime.
-  // Furthermore, until C++ interop is enabled, we cannot have a dependency on
-  // C++ code from pure Swift code.  If linked libraries are C++ based, they
-  // should properly link C++.  In the case of static linking, the user can
-  // explicitly specify the C++ runtime to link against.  This is particularly
-  // important for platforms like android where as it is a Linux platform, the
-  // default C++ runtime is `libstdc++` which is unsupported on the target but
-  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
-  // be present.  This results in linking the wrong version of libstdc++
-  // generating invalid binaries.  It is also possible to use different C++
-  // runtimes than the default C++ runtime for the platform (e.g. libc++ on
-  // Windows rather than msvcprt).  When C++ interop is enabled, we will need to
-  // surface this via a driver flag.  For now, opt for the simpler approach of
-  // just using `clang` and avoid a dependency on the C++ runtime.
-  const char *Clang = "clang";
+  const char *Clang =
```

Comment:
I think that this might want to use `clang-cl` to get all the additional defines that are needed for the Windows C++ runtime.  Selecting the driver could be a shared function in `Toolchain`.  Similar to Unix, could you rename this to `driver`?

Reply:
> I think that this might want to use `clang-cl` to get all the additional defines that are needed for the Windows C++ runtime.

Do you mean just for the C++ case or also for C? (`clang-cl` sounds like it's the C variant, i.e. analogous to `clang`; it there also a C++ variant of `clang-cl`?)

> Selecting the driver could be a shared function in `Toolchain`.

I understand the sentiment, but I don't really see a good way to do this.

Presumably, this would be some non-virtual function called, maybe, `getClangLinkerDriver()`? But if we want it to return `clang` on Unix and `clang-cl` on Windows, how would we get it do to that?

We could instead make the function virtual and provide different implementations for Unix and Windows, but now we have the same amount of code duplication that we have currently, and we have the additional question of how to implement this function for Darwin (which doesn't even use `clang` to link).

Am I misunderstanding? Did you have something different in mind?

> Similar to Unix, could you rename this to `driver`?

Done (renamed to `LinkerDriver`).

Reply:
Switching between `clang` and `clang-cl` would be based on `T.isOSWindows()` where `T` is `llvm::Triple`.  The Darwin case is something which needs to be addressed (both Jordan and I agreed on this but neither one of us got around to fixing :-() - it should switch over to `clang` as the linker driver, the original reasons for embedding the logic no longer apply as clang is so intricately tied into Swift.

Reply:
> Switching between `clang` and `clang-cl` would be based on `T.isOSWindows()` where `T` is `llvm::Triple`.

Shouldn't this be `--driver_mode=cl` rather than `clang-cl`? (See discussion above.)

This also makes this something that is easier to implement, because it's just something that can be added to `Arguments` locally in `WindowsToolChains.cpp` (i.e. the function that finds the driver doesn't need to inspect the triple).

> The Darwin case is something which needs to be addressed (both Jordan and I agreed on this but neither one of us got around to fixing :-() - it should switch over to `clang` as the linker driver, the original reasons for embedding the logic no longer apply as clang is so intricately tied into Swift.

Ah, got it.

For the time being, though, all I've done is to pull out the code for finding the linker driver into a common function in `ToolChain`. I'd like to keep this particular PR a behavior-preserving change. Followup PRs will then add `--driver_mode=cl` on Windows and switch Darwin to use Clang as the linker driver.

This will isolate the behavior changes into smaller PRs, which makes it easier to revert them if it turns out they have undesired consequences. WDYT?

Reply:
Sounds good to me.